architecture/design/pessimistic-locking-functional-spec.md

```diff
@@ -0,0 +1,661 @@
+## [Functional Spec] YSQL Pessimistic Locking
+
+
+## 1 Introduction
+
+To understand pessimistic locking behaviour in Pg, it is important to take note of some points about row-level locking and its interaction with DMLs.
+
+<span style="text-decoration:underline;">PostgreSQL supports 4 types of row-level locks -</span>
+
+
+
+1. Exclusive modes (i.e., only 1 txn can hold such a lock at any time) -
+    1. `FOR UPDATE` - takes exclusive lock on the whole row (blocks the shared locks as well)
+    2. `FOR NO KEY UPDATE` - takes exclusive lock on row but allows share lock on pk i.e., `FOR KEY SHARE`
+2. Share access modes
+    3. `FOR SHARE` - shared lock on full row i.e., blocks all exclusive locks
```

Comment:
Restart numbering. Ideally, just do all numbers as `1.` and let markdown figure out the sequencing. Indentation will sort out the sub-lists.

Reply:
done